### PR TITLE
Make integration test fully async

### PR DESCRIPTION
**Pull request by slang25**
Attempt to kill almost all blocking code, this removes all the occurences in the test projects. There are only 2 places left now:
- [Fluent API](https://github.com/justeat/JustSaying/search?q=GetAwaiter+filename%3AJustSayingFluently.cs&unscoped_q=GetAwaiter+filename%3AJustSayingFluently.cs), this we are well aware of
- ~~[Thottled.cs](https://github.com/justeat/JustSaying/blob/fa06e1b5f5ca35f11d734c9877141f26c77d369c/JustSaying/Messaging/MessageProcessingStrategies/Throttled.cs#L24), this one is tricky and could be hit under contention when at throttling limits, I have an idea on how to fix this in a non-breaking way. For another PR.~~

Something is going on with line-endings, I don't know if it's me, or if the code base needs all lines normalizing. I'll investigate.

### CODE REVIEW

**Comment by AnthonySteele** [comment]
Hm, we have a `.editorconfig`, but sometimes that's not enough and a compatible  `.gitattributes` file is also needed.

File: Directory.Build.props [hunk 1]
```diff
@@ -1,38 +1,40 @@
...
+    <PackageReleaseNotes>$(PackageProjectUrl)/releases</PackageReleaseNotes>
+    <PackageRequireLicenseAcceptance>false</PackageRequireLicenseAcceptance>
+    <PackageTags>aws,sns,sqs</PackageTags>
+    <PublicSign Condition=" '$(OS)' != 'Windows_NT' ">false</PublicSign>
+    <PublishRepositoryUrl>true</PublishRepositoryUrl>
+    <RepositoryType>git</RepositoryType>
+    <LangVersion>latest</LangVersion>
+    <RepositoryUrl>$(PackageProjectUrl).git</RepositoryUrl>
+    <!-- Temp until the JustBehave update is published -->
+    <RestoreSources>$(RestoreSources);$(MSBuildThisFileDirectory)local-packages/;https://api.nuget.org/v3/index.json</RestoreSources>
```
Comment (AnthonySteele) [inline Directory.Build.props]: This seems to be the main blocker. What is the situation with `JustBehave` ?

Comment (slang25) [reply Directory.Build.props]: The second part of the change we require here is approved (async Givens), and needs merging and releasing:
https://github.com/justeat/JustBehave/pull/31
Then we can reference the new beta. Possibly release a 2.0 and put it into "maintenance mode" ¯\_(ツ)_/¯

Comment (AnthonySteele) [reply Directory.Build.props]: Nice. It's double approved. Can you merge based on that?

Comment (slang25) [reply Directory.Build.props]: I have no permissions to JE repos other than JustSaying

Comment (AnthonySteele) [reply Directory.Build.props]: Oh well, merged now.

Comment (slang25) [reply Directory.Build.props]: Nice one 👍 Could someone push the packages at some point, it will be `2.0.0-beta-62`

**Comment by slang25** [comment]
Because we were hitting the fluent api code (that blocks) previously in the constructor, it was "harmless".

Now we are using async initialization, we deadlock the xunit SynchronizationContext.
There are lot's or hacky workarounds, I've opted here for disabling concurrent tests (which means xunit doesn't use it's special context). 🤠

**Comment by martincostello** [comment]
@slang25 I noticed you have an old [`fluent-refactor`](https://github.com/justeat/JustSaying/tree/fluent-refactor) lying around here. Do you still need it?

**Comment by slang25** [comment]
Any progress on a JustBehave release?

**Comment by adammorr** [comment]
@slang25 https://www.nuget.org/packages/JustBehave/2.0.0-beta-62 will be a thing once NuGet updates

**Comment by slang25** [comment]
Thanks @adammorr !

File: JustSaying.IntegrationTests/JustSaying.IntegrationTests.csproj [hunk 2]
```diff
@@ -18,9 +18,7 @@
     <PackageReference Include="Microsoft.NET.Test.Sdk" Version="15.9.0" />
     <PackageReference Include="Newtonsoft.Json" Version="11.0.2" />
     <PackageReference Include="StructureMap" Version="4.7.0" />
-    <PackageReference Include="xunit.runner.visualstudio" Version="2.4.1">
-      <PrivateAssets>all</PrivateAssets>
-      <IncludeAssets>runtime; build; native; contentfiles; analyzers</IncludeAssets>
-    </PackageReference>
+    <PackageReference Include="xunit.runner.visualstudio" Version="2.4.1" />
+    <PackageReference Include="xunit" Version="2.4.1" />
```
Comment (martincostello) [inline JustSaying.IntegrationTests/JustSaying.IntegrationTests.csproj]: Nit: these sort the opposite way around.

File: JustSaying.TestingFramework/TaskExtensions.cs [hunk 3]
```diff
@@ -1,14 +0,0 @@
-using System.ComponentModel;
```
Comment (martincostello) [inline JustSaying.TestingFramework/TaskExtensions.cs]: 🎉

File: JustSaying.UnitTests/JustSaying.UnitTests.csproj [hunk 4]
```diff
@@ -3,7 +3,7 @@
     <TargetFramework>netcoreapp2.1</TargetFramework>
   </PropertyGroup>
   <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Debug|AnyCPU'">
-    <NoWarn>$(NoWarn);CA2007;CA1051;CA1707;CA1034;CA1812;CA1307</NoWarn>
+    <NoWarn>$(NoWarn);CA2007;CA1051;CA1707;CA1034;CA1812;CA1307;CA1001</NoWarn>
```
Comment (martincostello) [inline JustSaying.UnitTests/JustSaying.UnitTests.csproj]: Nit: Can we sort these?

File: JustSaying.UnitTests/JustSayingFluentlyTestBase.cs [hunk 5]
```diff
@@ -1,8 +1,12 @@
 using System.Reflection;
+using System.Threading.Tasks;
 using JustSaying.AwsTools.QueueCreation;
 using JustBehave;
 using Microsoft.Extensions.Logging;
 using NSubstitute;
+using Xunit;
+
+[assembly: CollectionBehavior(DisableTestParallelization = true)]
```
Comment (martincostello) [inline JustSaying.UnitTests/JustSayingFluentlyTestBase.cs]: Can we put this in it's own `AssemblyInfo.cs` (or similar) file so it's easy to find?

File: JustSaying.TestingFramework/TaskExtensions.cs (hunk 3 above)
Comment (AnthonySteele) [reply JustSaying.TestingFramework/TaskExtensions.cs]: It's gone! Excellent.

**Comment by slang25** [comment]
I'll look at the failure over lunch

**Comment by slang25** [comment]
I've put the tests back to running sequentially now, there are too many areas that require exclusivity. Ready now @martincostello

**Comment by martincostello** [comment]
Go go gadget tests!

**Pull request merged**